ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,246 @@
+}
+
+/// Yields execution.
+pub fn yield_now(flags: YieldFlags) {
+    reschedule(&mut |local_rq| {
+        if flags != YieldFlags::Exit {
+            local_rq.update_current(flags.into());
+        }
+
+        let mut current_task_option = None;
+        if flags != YieldFlags::Yield {
+            current_task_option = local_rq.dequeue_current();
+        }
+        let current_task;
+        let mut current_task_inner = None;
+        if flags == YieldFlags::Wait {
+            current_task = current_task_option.clone().unwrap();
+            current_task_inner = Some(current_task.inner_exclusive_access());
+
+            if current_task_inner.as_ref().unwrap().task_status == TaskStatus::Runnable {
+                drop(current_task_inner);
+                local_rq.set_current(current_task_option);
+                return ReschedAction::DoNothing;
+            }
+        }
+
+        if let Some(next) = local_rq.pick_next_current() {
+            if flags == YieldFlags::Wait {
+                debug_assert_eq!(
+                    current_task_inner.as_ref().unwrap().task_status,
+                    TaskStatus::Sleepy
+                );
+                current_task_inner.unwrap().task_status = TaskStatus::Sleeping;
+            }
+            local_rq.set_current(Some(next.clone()));
+            ReschedAction::SwitchTo(next)
+        } else {
+            match flags {
+                YieldFlags::Exit => ReschedAction::Retry,
+                YieldFlags::Yield => ReschedAction::DoNothing,
+                YieldFlags::Wait => {
+                    drop(current_task_inner);
+                    local_rq.set_current(current_task_option);
+                    ReschedAction::DoNothing
+                }
+            }
+        }
+    })
+}
+
+/// Invokes a schedule.
+pub fn schedule() {
+    reschedule(&mut |local_rq| {
+        if !local_rq.should_preempt() {
+            return ReschedAction::DoNothing;
+        }
+        local_rq.set_should_preempt(false);
+
+        if let Some(next_current) = local_rq.pick_next_current() {
+            local_rq.set_current(Some(next_current.clone()));
+            ReschedAction::SwitchTo(next_current)
+        } else {
+            ReschedAction::DoNothing
+        }
+    });
+}
+
+/// Performs rescheduling.
+///
+/// This function accepts a closure whose result indicates its possible action.
+fn reschedule<F>(f: &mut F)
+where
+    F: FnMut(&mut dyn LocalRunQueue) -> ReschedAction,
+{
+    let mut next = None;
+    let mut flag = true;
+    while flag {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            match f(local_rq) {
+                ReschedAction::DoNothing => {
+                    flag = false;
+                }
+                ReschedAction::Retry => {}
+                ReschedAction::SwitchTo(next_task) => {
+                    next = Some(next_task);
+                    flag = false;
+                }
+            };
+        });
+    }
```

Comment:
The `flag` is redudant, right? You can write a loop and call `break` when you set `flag` as `false`.

Reply:
I'll fix this as advised